### ISSUE TITLE
"FutureWarning: elementwise comparison failed" makes it hard to check whether a value exists in an array

### ISSUE THREAD

**Issue by MarkMoulton** [numpy/numpy#6784]
I have innumerable places in my code where I test whether a value exists in an array when I don't know either the type of the value or the dtype of the array.  In previous numpy versions (I'm at 1.10.1), mismatched types return False.  Now they generate a FutureWarning.

``` Python
a = np.arange(5)

'x' in a
__main__:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
Out[78]: False
```

The same applies when comparing arrays elementwise:

``` Python
a = np.arange(3)

b = np.arange(3).astype(str)

a == b
__main__:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
Out[106]: False
```

I know the new behavior is deliberate, but I think practicality has been sacrificed for purity.

In the example above, it seems reasonable that `'x' in a` returns False, because being a string type automatically means 'x' cannot belong to an array of a different type.  It also makes sense that two arrays are not equal if they have different dtypes.

So how do I repair my code?  Do I add type checking for all value checks?  Do I convert the array to a list first?  Do I convert to an object dtype?  Do I use Try/Except?  I don't like any of these options and am having trouble thinking of a clean fix.  

For what it's worth, I view this change as a serious regression.

**Comment by charris** [comment]
Hmm, I'd say the function in the ndarray `sq_contains` slot is incorrectly implemented.
```
array_contains(PyArrayObject *self, PyObject *el)
{
    /* equivalent to (self == el).any() */

    PyObject *res;
    int ret;

    res = PyArray_EnsureAnyArray(PyObject_RichCompare((PyObject *)self,
                                                      el, Py_EQ));
    if (res == NULL) {
        return -1;
    }
    ret = array_any_nonzero((PyArrayObject *)res);
    Py_DECREF(res);
    return ret;
}
```
Looks like it might be good when the element wise comparisons take effect, but not at the moment.

**Comment by charris** [comment]
Although it may also be a bug in richcompare
```
In [14]: 'x' == arange(5)
Out[14]: False

In [15]: 1 == arange(5)
Out[15]: array([False,  True, False, False, False], dtype=bool)
```

**Comment by charris** [comment]
I'd guess that `array_compare` should return False if there is no common type.

**Comment by njsmith** [comment]
IMO it should return an appropriately shaped array of False if there's no common type.

It does sound like the bug is just that the warning is being emitted spuriously, and the code itself should be fine both before and after the change.

**Comment by MarkMoulton** [comment]
I think a decision was made that comparisons should result in Booleans of the same shape, implying that ```'x' in np.arange(5)``` would yield an array with five False values.  I sort of get the motivation but worry that it will result in awkward syntax and break a lot of code.  An expression like:

```Python
a = np.arange(5)
if 'x' in a:
print 'do this'
```
becomes something like
```Python
if np.all('x' in a == False):
print 'do this'
```
It sort of makes sense but makes for odd syntax.  I prefer a single Boolean False when too arrays have different types.  That Numpy managed like this for years without major problems suggests that the lack of consistency (returning a single Boolean vs an array of Booleans) is not a big deal, and might not be worth the hassle of breaking existing user code.

**Comment by njsmith** [comment]
Yes, the decision was made (many years ago!) that comparisons should return bool arrays of the appropriate shape, but that's for comparison operators like `==`. `in` is not a comparison operator and there's no plan to change it to return an array :-).

You're getting the warning because the implementation of `in` is calling `==` internally.

**Comment by MarkMoulton** [comment]
I think it goes beyond how "in" is implemented.  When ```==``` compares two arrays of different type the result has been, and still is except for the FutureWarning, a single Boolean False.
```Python
a = np.arange(5)
b = np.arange(5).astype(str)
a == b
__main__:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
Out[115]: False
```Python
As with the other example, I'd prefer the single Boolean return to a syntax like:
```if np.all((a == b) == False):
    print 'do this'

**Comment by seberg** [comment]
I don't quite understand, the spurrious warning should not be there, but aside from that the logic of `in` will not change. As to the second part:
```
if np.all(a == b):  # why the == True?
     print("do this")
```
The `np.all` can only be skipped if you already know the comparison is not possible due to incompatible dtypes, or if you have a 0-d array, in which case the result (basically) does not change and no FutureWarning is given. Do you need a way to find that the dtypes cannot be compared?

**Comment by MarkMoulton** [comment]
Let me make sure I understand.  In the special case where a, b, 'x' have incompatible types or dtypes, what will happen with the expressions ```a == b```  and ```'x' in a```?  Will they return an exception or an appropriately sized array with all False?  Will it be the same for ```==``` and ```in```?  Based on this thread, it doesn't sound like that's been decided.

Up until now, if they have incompatible types, ```a == b```  and ```'x' in a``` will only return False.  A lot of my code assumes that a "contains" or comparison with incompatible types will generate a False.  Therefore, when the change goes through it will break a lot of my code and until then generate a lot of FutureWarnings.  That's issue number 1.  And it's a big deal for me.

Issue number 2 is figuring out how to rewrite my code to be robust to contains and comparisons of arrays/scalars with incompatible types (I often have no way of knowing ahead of time what the types/dtypes of my inputs will be).  In the past, it was easy.  A simple ```if a == b: do this``` or ```if 'x' in a: do that``` handled all cases.  Now they won't.  

If the contains/comparison will generate an array of False, then a statement like ```if np.all((a == b) == False):  don't do this``` or ```if np.all('x' in a == False): don't do that``` should do the trick.

If the contains/comparison with incompatible types will throw an exception, then the rewrite gets nasty.  My options seem to be:  
1. check for compatibility of types prior to doing each comparison (this is straightforward when comparing arrays, a nightmare when using contains or comparing a Python scalar to an array)
2. put all comparisons in a try/except block (messy code, plus awkward when I have to catch warnings instead of exceptions)
3. change all arrays to lists or dtype=object

In all cases, the future change requires messier code than the status quo, and maybe buggier.  That's why I call the future change a serious regression.  Regardless, I need to know what the new behavior will be in order to make the appropriate fixes to my code.

Am I alone in this?

**Comment by njsmith** [comment]
Again, there is no plan to change `in`. The warning you are seeing is confusing and shouldn't be there.

For `==`, the current plan is that it should always return a boolean array of appropriate shape, regardless of whether the dtypes match. (Or an error if one of the arguments can't be converted to an array at all.)

However, I'm confused about why this is such a problem for you, because it's never been true that you can usefully write `if x == arr: ...`. In the past, that comparison could sometimes return False, but it could never return True. The possible return values were either False, or else an array that would then generate an error when passed to `if`.

The one case where I've seen this be an issue is when `arr` is sometimes not an array at all, e.g. for a kwarg that's either a string like `"auto"` or else an array of numbers. In this specific case it used to be possible to get away with writing `if mykwarg == "auto": ...`, and now one has to write `if isinstance(mykwarg, str) and mykwarg == "auto": ...`. But this is just the usual python idiom for handling type-variant arguments, and I think a reasonable price to pay to make array comparison follow consistent rules.

By the way, `not np.all((a == b) == False)` is probably much clearer when written as `np.any(a == b)` :-).

**Comment by MarkMoulton** [comment]
Thanks for the reply and for the ```np.any(a == b)``` tip.  I'm glad that the behavior of ```'x' in np.arange(5)``` will continue to return False.  And I understand the plan a little better.

I'm still unhappy about the plan to always return a boolean array of appropriate shape regardless of type compatibility:
1.  My code will stop working and it will take a lot of work for me to clean it up
2.  To be robust to type/dtype incompability, all comparisons written post-change won't work on older versions of Numpy.  All comparisons written pre-change won't work on newer versions of numpy. That's because a return of False cannot be treated as equal to an array of False.

Was this break in backwards compatibility really intended?

I wasn't thinking of ```if 'x' == arr: ...``` but of something like ```arr[arr == 'x']```, where the type of 'x' is incompatible with arr.  But since the comparison will return an array of False, here the future change works in our favor.  Of course, I still get all those FutureWarnings, which are a pain:
```Python
a = np.arange(5)
a == 'x'
__main__:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
Out[133]: False
```
But my big issue is losing the dtype robustness of ```a == b```.  I think that's going to cause a lot of problems.

**Comment by njsmith** [comment]
I think it would be helpful if you could provide some more concrete examples of what you're doing in your code that will break and that will require changes that break compatibility with older numpy's? (It would be fine if you want to take some time first to try updating some scripts and see how it works out in practice.) I've updated my personal code and it was pretty easy to do, plus the obvious fixes worked fine with older numpy versions too. But obviously YMMV. At the moment the problem is that I'm just not sure where exactly the issues are that you're worrying about -- the only concrete example you give in your comment is one where you argued yourself into preferring the new behavior :-).

**Comment by MarkMoulton** [comment]
Fair enough.  I'll go through my code and see what I can pull together.

In the meantime, the following snippet (in addition to the ```'x' in a``` example above) generates a FutureWarning.  Based on our discussion, I believe that the snippet actually should work in the future, contrary to what the FutureWarning implies.  If so, we at least have the issue of an over-zealous warning message.

```Python
def count_val(arr, val):
    ix = (arr == val) | (arr == str(val))
    return ix

print count_val(np.arange(4), 3)   # Generates FutureWarning
```

**Comment by njsmith** [comment]
Yes, that's correct that that code snippet will generate a FutureWarning in almost all cases, and that it will continue to work correctly in the future.

That particular case is not something we can really do anything about, unfortunately -- the intermediate values will in fact change. It happens you're using them in a way that will eventually produce the same result either way, but there's no way for numpy to know that. You can use the warnings module to just filter out the warning, if you want.

One of the major motivations for this change is to make string handling in numpy more reliable and consistent (right now it's really weird and hacky), so if you often write code like the above then it's not surprising that you're hitting the issue more often than most people. Hopefully this also means that you'll benefit more in the long run too :-)

(As a rule of thumb though, I'd generally recommend not doing automagic string coercion like in that snippet. There's a reason that python decided not to go down the same path as other scripting languages like tcl/perl/php and doesn't automatically coerce between e.g. numbers and strings -- doing this loses a lot of the value of Python's strong type system.)

**Comment by charris** [comment]
@njsmith Any reason to consider a fix for this? I'd like to remove the 1.10.2 milestone but want to make sure it is the right thing to do.

**Comment by njsmith** [comment]
I guess it might be nice to suppress the warning for `string in array` if there's a clean way to do it, but let's not hold up 1.10.2 waiting for this.

**Comment by charris** [comment]
OK, I'll go ahead with 1.10.2rc2 and if a simple fix turns up in the next week we can add it for the final. Although fooling with the C code just before a release makes me a bit nervous...

**Comment by MarkMoulton** [comment]
Here's a quick report on the places in my code where a FutureWarning gets called.  They are all triggered by incompatible types/dtypes and come up in expressions of the sort that appear in the following toy examples:
```Python
import numpy as np

def get_ix(arr, x):
    ix = np.where(arr == x)
    return ix

def x_in_arr(arr, x):
    if x in arr:
        return True
    else:
        return False

def match_arrays_now(a, b):
    if np.all(a == b):
        return 'matched'
    else:
        return 'no match'

def match_arrays_future(a, b):
    
    # Case: a == b, incompatible types, yielding f
    f = np.array([False, False, False])
    if np.all(f):
        return 'matched'
    else:
        return 'no match'

def not_matched_arrays_now(a, b):
    if np.any(a != b):
        return 'not matched'
    else:
        return 'match'
    
def not_matched_arrays_future(a, b):
    
    # Case: a != b, incompatible types, yielding t
    t = np.array([True, True, True])
    if np.any(t):
        return 'not matched'
    else:
        return 'match'

# Quickie tests
a = np.arange(4)
b = a.astype('S2')

#get_ix(a, 'x')  # empty array, good, but FutureWarning
#get_ix(a, 3)    # index, good, no warning
#x_in_arr(a, 'x')  # False, good, but FutureWarning
#x_in_arr()=', x_in_arr(a, 3)  # True, good, no warning

#match_arrays_now(a, b) # 'no match', good, but FutureWarning
#match_arrays_now(a, np.copy(a)) # 'matched', good, no warning
#match_arrays_future(a, b)  # 'no match', good (warning N/A)
#not_matched_arrays_now(a, b)  # 'not matched', good, but FutureWarning
#not_matched_arrays_future(a, b) # 'not matched', good, (warning N/A)
```
The good news isn't that it doesn't look like the future changes will break my code, or code that uses these idioms.  However, a lot of code that should continue to work is currently generating FutureWarnings and may cause people like me to get paranoid and try to fix what ain't broken.

**Comment by njsmith** [comment]
For the all(a == b) and any(a != b) cases you might want to look at np.array_equal.

**Comment by xiaoleihuang** [comment]
Interesting, I also encountered such a problem when I used 
`skf = StratifiedKFold(data_labels, n_folds=5)` 
`for train_index, test_index in skf:`

It might be caused by sklearn precision_recall_fscore_support function as well, because I also used it as below:
`precision, recall, f1, support = precision_recall_fscore_support(
                    tweets_labels[test_index], y_predict, average=None, labels=['0', '1'])`

**Comment by DStauffman** [comment]
It looks like this also impacts comparisons with timedelta64 objects and gives very unpredictable results.

    # np.__version__ == '0.19.1'
    np.timedelta64(0, 'ns') == 0  # raises warning and returns False
    np.timedelta64(0, 'ns') == np.array(0)  # raises warning and returns False
    np.timedelta64(0, 'ns').astype(np.int64) == np.array(0)  # no warning, returns True
    np.timedelta64(0) == 0  # no warning, returns True
    np.timedelta64(0) == np.array(0)  # no warning, returns True

**Comment by ejhyer** [comment]
I just made a change that caused this behavior to change, from numpy 1.20.3 to 1.21.4. 
    `numpy_float_arr1 == numpy_float_arr2` previously returned an array of Boolean, now returns a Warning and a single Boolean value.

**Comment by seberg** [comment]
@ejhyer could you give a few more details, about what you are doing?  I am not aware of any serious changes recently.  The only thing I can think of would be if you now run into a MemoryError, and that ends up replaced (which is not good and maybe we should prioritize fixing, independent of the general warning.).  And somehow on 1.20.x you got off lucky and did not get a memory error?!

**Comment by ejhyer** [comment]
Hi,

 

PEBKAC as usual-- we'll delete the exchange, with your consent. Sorry to waste your time.

 

Arrays did not agree, I was getting them from xarray.values and the .size was the same but the .shape was different.

 

I ran the code against new data that somehow had the dimensions ordered differently, so an error was the appropriate response. Had I used .values.flat, I would not have gotten an error from numpy, but I would have gotten a very wrong result.

 

Thanks for your quick response, sorry for wasting your time.

 

--Edward H.
